pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java

```diff
@@ -178,92 +190,248 @@ public CompletableFuture<Void> deleteTopicPoliciesAsync(TopicName topicName,
+     * @return a CompletableFuture that completes when the message ID has been read by the reader
+     */
+    private CompletableFuture<Void> untilMessageIdHasBeenRead(NamespaceName namespaceObject, MessageId messageId) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        getMessageHandlerTracker(namespaceObject).addPendingFuture((MessageIdAdv) messageId, future);
+        return future;
+    }
+
+    private TopicPolicyMessageHandlerTracker getMessageHandlerTracker(NamespaceName namespaceObject) {
+        return topicPolicyMessageHandlerTrackers.computeIfAbsent(namespaceObject,
+                ns -> new TopicPolicyMessageHandlerTracker());
+    }
+
+    private record PendingMessageFuture(MessageId messageId, CompletableFuture<Void> future)
+            implements Comparable<PendingMessageFuture> {
+        @Override
+        public int compareTo(PendingMessageFuture o) {
+            return messageId.compareTo(o.messageId);
+        }
+    }
+
+    /**
+     * This tracks the last handled message IDs for each partition of the topic policies topic and
+     * pending futures for topic policy messages. Each namespace has its own tracker instance since
+     * this is tracking the per-namespace __change_events topic.
+     * The purpose for this tracker is to ensure that write operations on topic policies don't complete before the topic
+     * policies message has been read by the reader and effective.
+     */
+    private static class TopicPolicyMessageHandlerTracker implements AutoCloseable {
+        private List<MessageIdAdv> lastHandledMessageIds = new ArrayList<>();
+        private List<PriorityQueue<PendingMessageFuture>> pendingFutures = new ArrayList<>();
+        private boolean closed = false;
+
+        /**
+         * Called after a message ID has been handled by the reader.
+         * This will update the last handled message ID for the partition and complete any pending futures that are
+         * registered to the handled message ID or before it.
+         * @param messageId the message ID that has been handled
+         */
+        public synchronized void handleMessageId(MessageIdAdv messageId) {
+            if (closed) {
+                return;
+            }
+            int partitionIndex = messageId.getPartitionIndex();
+            if (partitionIndex < 0) {
+                partitionIndex = 0;
+            }
+            while (lastHandledMessageIds.size() <= partitionIndex) {
+                lastHandledMessageIds.add(null);
+            }
+            lastHandledMessageIds.set(partitionIndex, messageId);
+            if (pendingFutures.size() > partitionIndex) {
+                PriorityQueue<PendingMessageFuture> pq = pendingFutures.get(partitionIndex);
+                while (!pq.isEmpty() && pq.peek().messageId.compareTo(messageId) <= 0) {
+                    PendingMessageFuture pendingFuture = pq.poll();
+                    pendingFuture.future.complete(null);
```

Comment:
Completing the future in a synchronized block is dangerous because all callbacks registered on the future will be executed when holding the lock.

It would be better to use a separated thread to complete the pending future

Reply:
I've run the `TopicPoliciesUpdateTest` and found the contention happened between `broker-client-shared-internal-executor` (client I/O thread) and `pulsar-io` (broker I/O thread). i.e. when the future is completed in `handleMessageId`, the `pulsar-io` threads might be blocked by the callbacks.

The callbacks should only be registered by `updateTopicPoliciesAsync` and `deleteTopicPoliciesAsync`, which are only. No thread switching is okay, because all these method calls are only called in `PersistentTopicsBase`, where the callback just logs a message and call `asyncResponse.resume`.